src/cmap/handshake/client_metadata.ts

```diff
@@ -152,8 +153,62 @@ export function makeClientMetadata(options: MakeClientMetadataOptions): ClientMe
       }
     }
   }
+  return metadataDocument.toObject() as ClientMetadata;
+}
+
+let isDocker: boolean;
+let dockerPromise: Promise<void>;
+/** @internal */
+async function getContainerMetadata() {
+  const containerMetadata: Record<string, any> = {};
+  if (isDocker == null) {
+    dockerPromise ??= fs.access('/.dockerenv');
+    try {
+      await dockerPromise;
+      isDocker = true;
+    } catch {
+      isDocker = false;
+    }
+  }
```

Comment:
there's a million ways to skin a cat - but just because I think it's a bit cleaner not to cache to variables:

```suggestion
  dockerPromise ??= fs.access('/.dockerenv').then(() => true, () => false);
  const isDocker = await dockerPromise;
```

Reply:
`fs.access` has a return type of Promise<void> not Promise<boolean> so I'm more comfortable with the current implementation as it stands.

Reply:
> fs.access has a return type of Promise not Promise

could you clarify?

Reply:
(The angle brackets were interpreted as html, fixed)

Reply:
I'm fine leaving it but I don't really understand the objection.  It resolves to undefined if the file is accessible and rejects otherwise - that's why we map it to a boolean

Reply:
Just pointing out, there _is_ a real difference between the current code and @baileympearson's suggestion; if you call this function twice in a row without waiting for the result, i.e. `getContainerMetadata(); getContainerMetadata();`, then the current code will call `fs.access` twice and the suggestion here would only call it once

Reply:
I see, I'm good to switch to Bailey's implementation.

Reply:
@addaleax I don't think so - because we use the `??=` operator.  We'll enter the if-statement twice, but after the first call the promise will no longer be undefined so we'll only call `fs.access` once.  Right?

Reply:
ah, right – I did miss that the original code actually did two checks, one on `isDocker` and then another one on `dockerPromise`. ignore me then :)

Reply:
I'm still okay with switching to Bailey's suggestion since the code is smaller.

Reply:
https://jira.mongodb.org/browse/NODE-5916

The ticket above intends to make this kind of code discouraged, of course, we can always exempt cases where it makes sense but I'll add that this creates a new promise rather than using the one already returned by `fs.access()`.

Fine by me to leave it as is for this PR, but it may come into question when that ticket has to correct the lint errors.

Reply:
good to know!